### ISSUE TITLE
3.13 Deprecated decorator has issues with decorator stacking

### ISSUE THREAD

**Issue by max-muoto** [microsoft/pyright#7732]
**Describe the bug**
When attempting to use the new `deprecated` decorator that's [being released in 3.13](https://peps.python.org/pep-0702/) (through typing extensions), combining another decorator with causes the Pyright warning to be lost.

**Code or Screenshots**
Take this simple, no matter if the deprecated decorator comes first or last, usage doesn't trigger a Pyright error.

```python
from typing_extensions import deprecated
from functools import cache


@cache
@deprecated("Don't use this!")
def foo() -> None: ...

foo() # no error!
```
[Pyright Playground
](https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmB9ApgDxllAZyTCKiQgTBBigBMsEQsBjAQ3zoChRIpgArihYwwYADaFylarXYsAFli4qAAvKVdVDJqw5Y6ACgBEAEVIByWgMJZYCpIQCExgJRcGwfmMOuoAWgA%2BKAA5UiwALigAOliVYB8-AGIoFDAoLBBwECcgA)

As a note the example [PEP 702](https://peps.python.org/pep-0702/) that includes examples of decorator stacking with properties, setters, and overloads works as expected.

Pyright 1.1.359

**Comment by erictraut** [comment]
Pyright is working correctly here, so I don't consider this a pyright bug.

This is a problem with the `functools.cache` decorator, which does not preserve information about the decorated function's input signature. This has already been reported to the maintainers of typeshed, and there is an ongoing discussion about how to fix the definition of `functools.cache` in the stubs. Until that happens, the input signature of a function decorated with `@cache` — along with its docstring and deprecation information — will not be preserved.

**Comment by max-muoto** [comment]
I was using `cache` as an example, but I probably should of used this, as it was closer to what I was seeing in reality. In this case, it would be great to avoid deprecating the container itself, but there's likely not a way around this I suppose.

**Comment by erictraut** [comment]
Ah, that's a different case. Yes, that's also working as intended, although an argument could be made that the behavior should be changed here.

When a ParamSpec "captures" the signature of a function, pyright records not only the input signature (i.e. its list of parameters), it also records some additional metadata including the function's docstring, deprecation info, and some internal flags indicating whether the function is an instance or class method. When the solved ParamSpec is applied to the return type of the decorator, pyright needs to decide whether to retain this additional metadata or discard it.

If a decorator returns a `Callable[P, R]`, it's pretty likely that it's simply returning the same function that it was passed, so pyright retains the metadata in this case.

If a decorator returns `Callable[Concatenate[<something>, P], R]`, it's a little less clear, but it's still likely that the decorator is returning a wrapped function, so pyright retains the metadata in this case too.

If the decorator returns an instance of a wrapper class parameterized by `P` (such as `AsyncTask[P, R]` in your example), it's unclear whether this metadata should be retained. A ParamSpec can be used in many ways and for many reasons within a class definition. One could argue that this metadata (or some subset of it) should be retained in this case, but there are good arguments to the contrary. In particular, I think it would be a problem to retain the flags that indicate whether it's an instance or class method. It would perhaps be safe to propagate the docstring and deprecation information, but it feels inconsistent to propagate some metadata and not all of it.

The typing spec is currently silent about whether this metadata should be retained (and under what conditions) when a signature is captured by a ParamSpec. In fact, the typing spec doesn't even recognize that there is metadata associated with a ParamSpec. This is maybe something we should attempt to standardize.

**Comment by max-muoto** [comment]
> Ah, that's a different case. Yes, that's also working as intended, although an argument could be made that the behavior should be changed here.
> 
> When a ParamSpec "captures" the signature of a function, pyright records not only the input signature (i.e. its list of parameters), it also records some additional metadata including the function's docstring, deprecation info, and some internal flags indicating whether the function is an instance or class method. When the solved ParamSpec is applied to the return type of the decorator, pyright needs to decide whether to retain this additional metadata or discard it.
> 
> If a decorator returns a `Callable[P, R]`, it's pretty likely that it's simply returning the same function that it was passed, so pyright retains the metadata in this case.
> 
> If a decorator returns `Callable[Concatenate[<something>, P], R]`, it's a little less clear, but it's still likely that the decorator is returning a wrapped function, so pyright retains the metadata in this case too.
> 
> If the decorator returns an instance of a wrapper class parameterized by `P` (such as `AsyncTask[P, R]` in your example), it's unclear whether this metadata should be retained. A ParamSpec can be used in many ways and for many reasons within a class definition. One could argue that this metadata (or some subset of it) should be retained in this case, but there are good arguments to the contrary. In particular, I think it would be a problem to retain the flags that indicate whether it's an instance or class method. It would perhaps be safe to propagate the docstring and deprecation information, but it feels inconsistent to propagate some metadata and not all of it.
> 
> The typing spec is currently silent about whether this metadata should be retained (and under what conditions) when a signature is captured by a ParamSpec. In fact, the typing spec doesn't even recognize that there is metadata associated with a ParamSpec. This is maybe something we should attempt to standardize.

Makes sense, really appreciate the clarification. To maybe add some light, I feel this pattern is fairly common for task based libraries, that's how [Celery](https://github.com/celery/celery) is stubbed out for us, and so we have a lot of these async tasks throughout our codebase, so I think it would be useful long-term if there was someway to indicated the returned-class as deprecated. Perhaps long-term there could be a qualifier to indicate such, if the metadata isn't going to be passed along via `P`:

```python
def make_task(f: Callable[P, R]) -> Deprecated[AsyncTask[P, R]]:
    container = AsyncTask(f)
    return container
```

However, it seems this idea was considered in the PEP:

> For deprecating module-level constants, object attributes, and function parameters, a Deprecated[type, message] type modifier, similar to Annotated could be added. However, this would create a new place in the type system where strings are just strings, not forward references, complicating the implementation of type checkers. In addition, my data show that this feature is not commonly needed.

I would see it fairly useful, for both this case, and module-level constants, but nonetheless.

**Comment by max-muoto** [comment]
CC: @JelleZijlstra if you have any thoughts here.

**Comment by JelleZijlstra** [comment]
I would encourage a follow-up PEP to add a way to deprecate module-level constants and other objects that PEP 702 didn't support.

I don't understand your example, `def make_task(f: Callable[P, R]) -> Deprecated[AsyncTask[P, R]]:`, though. If you want to deprecate all uses of `make_task()`, why not apply the `@deprecated` decorator to the `make_task` function? If instead you want to make the result deprecated only if the argument was deprecated, you need the type checker to carry over the deprecation information with the ParamSpec, as Eric wrote above.

**Comment by erictraut** [comment]
I dug into this further, and I think I've convinced myself that it's safe to pass the docstring and deprecation information through a ParamSpec. I've added support for this pattern. This will be included in the next release.

**Comment by max-muoto** [comment]
> I would encourage a follow-up PEP to add a way to deprecate module-level constants and other objects that PEP 702 didn't support.
> 
> I don't understand your example, `def make_task(f: Callable[P, R]) -> Deprecated[AsyncTask[P, R]]:`, though. If you want to deprecate all uses of `make_task()`, why not apply the `@deprecated` decorator to the `make_task` function? If instead you want to make the result deprecated only if the argument was deprecated, you need the type checker to carry over the deprecation information with the ParamSpec, as Eric wrote above.

>>> `@deprecated` decorator to the `make_task` function?

Good point, probably not a use for that example I showed off. I think this is a better example to show off the use-case. For example if want to define your own custom deprecation decorator. As you know, a lot of libraries maintain their own custom deprecation decorators, some of them with custom logic for fetching the stack-level/creating the warning message that you can't get with the native one. 


```python
def custom_deprecator(
    message: str, version: str,
) -> Callable[[Callable[P, T]], Deprecated[Callable[P, T]]]:
    def decorate(function: Callable[P, T]) -> Deprecated[Callable[P, T]]:
        @wraps(function)
        def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
            warnings.warn(f"{function.__name__} is deprecated in version {version}. {message}", DeprecationWarning)
            return function(*args, **kwargs)

        return wrapper

    return decorate
```

It could be helpful to tweak the call signature based on your project's specific use-cases.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.360, which I just published.
